File: test/cpp/gpu/test_multi_stream_sync.cpp
#include <nexus.h>

#include <cstdlib>
#include <fstream>
#include <iostream>
#include <numeric>

std::vector<std::string_view> nexusArgs;

int main(int argc, char **argv) {
  if (argc < 3) {
    std::cout << "Usage: " << argv[0] << " <kernel_lib> <kernel_name>"
              << std::endl;
    return 1;
  }
  std::string kernel_lib = argv[1];
  std::string kernel_name = argv[2];
  auto sys = nexus::getSystem();
  auto runtimes = sys.getRuntimes();

  // Find the first runtime with a valid device
  nexus::Runtime rt;
  nexus::Device dev0;

  for (auto runtime : runtimes) {
    auto type = runtime.getProp<std::string>(NP_Type);
    if (type == "gpu") {
      auto devices = runtime.getDevices();
      if (!devices.empty()) {
        rt = runtime;
        dev0 = devices[0];
        break;
      }
    }
  }

  if (!rt || !dev0) {
    std::cout << "No runtime with valid devices found" << std::endl;
    return 1;
  }

  auto count = rt.getDevices().size();

  std::cout << "RUNTIME: " << rt.getProp<std::string>(NP_Name) << " - " << count
            << std::endl;

  for (int i = 0; i < count; ++i) {
    auto dev = rt.getDevice(i);
    std::cout << "  Device: " << dev.getProp<std::string>(NP_Name) << " - "
              << dev.getProp<std::string>(NP_Architecture) << std::endl;
  }
  std::vector<float> vecA(1024, 1.0);
  std::vector<float> vecB(1024, 2.0);
  std::vector<float> vecResult_GPU(1024, 0.0);  // For GPU result

  size_t size = 1024 * sizeof(float);

  auto nlib = dev0.createLibrary(kernel_lib);

  if (!nlib) {
    std::cout << "Failed to load library: " << kernel_lib << std::endl;
    return 1;
  }

  auto kern = nlib.getKernel(kernel_name);
  if (!kern) {
    std::cout << "Failed to get kernel: " << kernel_name << std::endl;
    return 1;
  }

  auto buf0 = dev0.createBuffer(size, vecA.data());
  auto buf1 = dev0.createBuffer(size, vecB.data());
  auto buf2 = dev0.createBuffer(size, vecResult_GPU.data());
  auto buf3 = dev0.createBuffer(size, vecResult_GPU.data());

  auto stream0 = dev0.createStream();
  auto stream1 = dev0.createStream();

  auto evFinal = dev0.createEvent();
  auto ev0 = dev0.createEvent();

  // Stream 0
  auto sched0 = dev0.createSchedule();

  auto cmd0 = sched0.createCommand(kern);
  cmd0.setArgument(0, buf0);
  cmd0.setArgument(1, buf1);
  cmd0.setArgument(2, buf2);

  cmd0.finalize(32, 32);

  sched0.createSignalCommand(ev0);

  // Stream 1
  auto sched1 = dev0.createSchedule();

  sched1.createWaitCommand(ev0);

  auto cmd1 = sched1.createCommand(kern);
  cmd1.setArgument(0, buf0);
  cmd1.setArgument(1, buf2);
  cmd1.setArgument(2, buf3);

  cmd1.finalize(32, 32);

  sched1.createSignalCommand(evFinal);

  // Run streams -- order is important for HIP events :-(
  sched0.run(stream0, false);
  sched1.run(stream1, false);

  evFinal.wait();

  buf3.copy(vecResult_GPU.data());

  int i = 0;
  for (auto v : vecResult_GPU) {
    if (v != 4.0) {
      std::cout << "Fail: result[" << i << "] = " << v << std::endl;
    }
    ++i;
  }
  std::cout << "Test PASSED" << std::endl;

  return 0;
}

File: test/pynexus/test_multi_stream_sync.py
import numpy
import torch

import nexus

buf0 = torch.ones(1024)
buf1 = torch.ones(1024)
res1 = torch.zeros(1024)
res2 = torch.zeros(1024)

for rt in nexus.get_runtimes():
  if (rt.get_devices().size()):
    dev = rt.get_devices()[0]
    if (dev.get_property_str(nexus.property.Type) == "gpu"):
      break

nb0 = dev.create_buffer(buf0)
nb1 = dev.create_buffer(buf1)
nb2 = dev.create_buffer(res1)
nb3 = dev.create_buffer(res2)

lib = dev.load_library("kernel.so")
kern = lib.get_kernel('add_vectors')

evf = dev.create_event()

stream0 = dev.create_stream()
stream1 = dev.create_stream()

# schedule 0
sched0 = dev.create_schedule()

cmd = sched0.create_command(kern)
cmd.set_arg(0, nb0)
cmd.set_arg(1, nb1)
cmd.set_arg(2, nb2)
cmd.finalize(32, 1024)

ev0 = sched0.create_signal()

# schedule 1
sched1 = dev.create_schedule()

sched1.create_wait(ev0.get_event())

cmd1 = sched1.create_command(kern)
cmd1.set_arg(0, nb0)
cmd1.set_arg(1, nb2)
cmd1.set_arg(2, nb3)
cmd1.finalize(32, 1024)

sched1.create_signal(evf)

# Run it
sched1.run(stream1, False)
sched0.run(stream0, False)

evf.wait()

nb3.copy(res2)

print(res2)
